R/read-gs.R

```diff
@@ -18,14 +19,225 @@
+  grades <- grades |>
+    dplyr::filter(!is.na(Student)) |>
+    dplyr::mutate_at(assignments, function(x) {
+        dplyr::if_else(x == "EX", NA, x) |>
+        as.numeric()
+      })
+  
+ 
+  # now reconfigure the points possible info
+  max_points <- dplyr::filter(grades, Student %in% c( "    Points Possible", "Points Possible")) |>
+    dplyr::select(all_of(assignments)) |>
+    dplyr::slice(rep.int(1, nrow(grades))) |>
+    dplyr::rename_with(
+        function(x){
+          paste0(x, " - Max Points")
+        }
+    )
+  
+  
+  # remove unneeded cols, add max_point column, and remove rows that are not students
+  grades <- grades |>
+    dplyr::select(c(all_of(assignments), "ID", "Student", 
+                    "SIS User ID", "Section")) |>
+    dplyr::bind_cols(max_points) |>
+    dplyr::filter(!(Student %in% c("    Points Possible", "Student, Test", "Points Possible", NA)))
+  
+  # add columns for submission time and lateness
+  
+  sub_cols <- paste0(assignments, " - Submission Time")
+  
+  late_cols <- paste0(assignments, " - Lateness (H:M:S)")
+  
+  grades[sub_cols] <- NA
+  
+  grades[late_cols] <- lubridate::make_difftime(NA)
+  
+  # extract student name data
+  
+  grades$`First Name` <- (stringr::str_match(grades$Student, 
+                                             ".+,\\s(.+)"))[, 2]
+  
+  grades$`Last Name` <- (stringr::str_match(grades$Student, 
+                                            "(.+),.+"))[, 2]
+  
+  # change names to match Gradescope convention
+  grades <- grades |>
+    dplyr::rename(SID = `SIS User ID`,
+                  Sections = Section) |>
+    # reorder columns to match Gradescope
+    dplyr::select(c("First Name", "Last Name", "SID", "Sections",
+                    as.vector(
+                            mapply(c,
+                                   assignments, 
+                                    paste0(assignments, " - Max Points"), 
+                                    sub_cols,
+                                    late_cols)
+                              )
+                    )
+                  ) |>
+    # remove "UID:" moniker and make SID numeric
+    dplyr::mutate_at("SID", function(x){
+      stringr::str_replace(x, "^UID:", "") |> 
+        as.numeric()
+    })
+  
+  attr(grades, "source") <- "Canvas"
+  
+  check_data_format(grades)
+  
 }
 
-#' Check Formatting of Gradescope Data
+#' Predict the source of grade dataframe.
+#'
+#' This function performs the auto-determination of where the grades dataframe was sourced.
+#' @param grades_df This is the input Dataframe
+#' @returns A string "Canvas", "Gradescope", or "Unrecognized" of the determination.
+#' @export
+determine_grade_source <- function(grades_df){
+  
+  columns <- names(grades_df)
+  
+ if (all(c("Student", "ID", "SIS User ID", "SIS Login ID") %in% columns)){
+   return("Canvas")
+ } else if (all(c("First Name", "Last Name", "SID", "Email") %in% columns)){
```

Comment:
I don't think Gradescope always had "First Name" and "Last Name" because I remember seeing just a "Name" column (possibly for smaller courses)... it may be better to use "SIS User ID" + "SIS Login ID" to check for "Canvas" and use similar logic to 'get_assignments()' to determine if it's Gradescope

Reply:
Also, check for Gradescope first, and then Canvas because I think assignment names formatting are more robust than checking id cols, so switch up the order of the if statements too